Ports paradise's modular buildmode

marlyn-x86 opened the pull request:
Whoops, let this sit too long. I've squashed and rebased to fix the conflicts. Repost of #39598
:cl: Crazylemon
admin: "Fill" Buildmode, "Area Edit" Buildmode, "Boom" Buildmode
refactor: Buildmode is now in modules instead of defining new modes by extra branches of switch statements.
imageadd: Switches buildmode button icons to something with a slightly more refined background.
admin: Buildmode now switches between modes using a discrete switcher, instead of by repeatedly clicking.
/:cl:
Ports ParadiseSS13/Paradise#9291

Features it includes:
+ Datumized buildmode sub-modes, to make the system more extensible.
+ A more user-friendly mode and dir-switching system.
+ The following buildmodes:
    + "Fill" Buildmode
    + "Area Edit" Buildmode
    + "Boom" Buildmode

Things I did not port due to wanting to get the simple stuff through first:
+ ["Atmos" Buildmode](https://github.com/ParadiseSS13/Paradise/blob/8edd8676be47a1f26b57caf508f108cb301da800/code/modules/buildmode/submodes/atmos.dm)
+ ["Map Saving" Buildmode](https://github.com/ParadiseSS13/Paradise/blob/8edd8676be47a1f26b57caf508f108cb301da800/code/modules/buildmode/submodes/save.dm)
+ ["Door Link" Buildmode](https://github.com/ParadiseSS13/Paradise/blob/8edd8676be47a1f26b57caf508f108cb301da800/code/modules/buildmode/submodes/link.dm)

Pictures of the new mode switching system:

<img width="277" alt="tgstation_bm_modeswitch" src="https://user-images.githubusercontent.com/5661700/43790066-6a43b254-9a26-11e8-88fb-a6ca6ae106a6.png">

<img width="258" alt="tgstation_bm_dirswitch" src="https://user-images.githubusercontent.com/5661700/43790065-6a18ba2c-9a26-11e8-9177-82f2604d4797.png">

marlyn-x86 committed:
Ports paradise's modular buildmode

Crazylemon64 commented on code/modules/buildmode/buildmode.dm:
```diff
@@ -0,0 +1,161 @@
+	// since these will get wiped upon login
+	holder.screen += buttons
+
+/datum/buildmode/proc/post_logout()
```
@AnturK I created it as I added callbacks to the `player_details` datum, and did not actually find a use for it. I'll remove it now.

Crazylemon64 replied:
Consequently, should I remove the entire `player_details` logout (not login) callback system? After this removal, nothing will utilize it.

marlyn-x86 committed:
Axes logout callback of buildmode

AnturK reviewed (changes requested):
Some minor stuff and this should be good to go.

marlyn-x86 committed:
AnturK's changes

- Code formatting
- Gives mob to throw args
- Selection overlays are on ABOVE_LIGHTING_PLANE to be visible on space

Other:
- Doubleclicks of observers and AI no longer "double-tap" click
intercepts

Crazylemon64 commented:
@AnturK Changes made
